File: sandlib/src/main/java/jomeslu/com/sandpic/SandPic.java
/*
 *  Copyright 2017 jomeslu (luzhensheng72@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jomeslu.com.sandpic;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.util.Log;

/**
 * Created by jomeslu on 17-3-28.
 */
public class SandPic {
    private static SandPic ourInstance = new SandPic();

    public static SandPic getInstance() {
        return ourInstance;
    }

    private SandPic() {
    }

    private static native int[] generate(int[] pixels, int width, int height, int threshold, int ponitNum);


    /**
     * 像素点转换
     * @param bitmap
     * @param threshold
     * @param ponitNum
     * @return
     */
    public Bitmap tramform(Bitmap bitmap,int threshold,int ponitNum ){
        int width =  bitmap.getWidth();
        int height = bitmap.getHeight();
        Bitmap newImage = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(newImage);
        Paint paint = new Paint();
        paint.setAntiAlias(false);
        paint.setStyle(Paint.Style.STROKE);
        int pixels[] = new int [width*height];
        bitmap.getPixels(pixels,0,width,0,0,width,height);
        int[] generate = generate(pixels, width, height, threshold, ponitNum);
        for (int i = 0, n = generate.length; i + 1 < n; i += 2) {
                int x = generate[i]>0? generate[i]:0;
                int y = generate[i+1] >0?generate[i+1]:0 ;
                int color = bitmap.getPixel(x,y);
                paint.setColor(color);
                canvas.drawCircle(x, y, 1, paint);
            }

        return newImage;
    }
    static {
        System.loadLibrary("sand-lib");
    }
}

File: sand-kotlin/src/main/java/com/mimobile/sandkotlin/SandPic.kt
package com.mimobile.sandkotlin

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint

/**
 * Created by jomeslu on 18-12-1.
 */
class SandPic private constructor() {
    private object Holder {
        val INSTANCE = SandPic()
    }

    private external fun generate(pixels: IntArray, width: Int, height: Int, threshold: Int, ponitNum: Int): IntArray

    fun tramform(bitmap: Bitmap, threshold: Int, ponitNum: Int): Bitmap {
        val width = bitmap.width
        val height = bitmap.height
        val newImage = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(newImage)
        val paint = Paint()
        paint.isAntiAlias = false
        paint.style = Paint.Style.STROKE
        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)
        val generate = generate(pixels, width, height, threshold, ponitNum)
        var i = 0
        val n = generate.size
        while (i + 1 < n) {
            val x = if (generate[i] > 0) generate[i] else 0
            val y = if (generate[i + 1] > 0) generate[i + 1] else 0
            val color = bitmap.getPixel(x, y)
            paint.color = color
            canvas.drawCircle(x.toFloat(), y.toFloat(), 1f, paint)
            i += 2
        }

        return newImage
    }

    companion object {
        // Used to load the 'native-lib' library on application startup.
        init {
            System.loadLibrary("sand-lib")
        }

        val instance: SandPic by lazy { Holder.INSTANCE }
    }

    var b: String? = null
}